pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,382 @@
+// to all events.
+type eventData struct {
+	// host specifies the host which the sender is sending to.
+	host string
+	// bytes represents the number of bytes affected by this event. It is
+	// not known for eventTypeFlushed.
+	bytes int
+	// count specfies the number of payloads that this events refers to.
+	count int
+	// duration specifies the time it took to complete this event. It
+	// is set for eventTypeFlushed, eventTypeSent, eventTypeRetry and
+	// eventTypeFailed.
+	duration time.Duration
+	// err specifies the error that may have occurred on events like
+	// eventTypeRetry and eventTypeFailed.
+	err error
+}
+
+// eventRecorder implementations are able to take note of events happening in
+// the sender.
+type eventRecorder interface {
+	// recordEvent notifies that event t has happened, passing details about
+	// the event in data.
+	recordEvent(t eventType, data *eventData)
+}
+
+// maxQueueSize specifies the maximum allowed queue size. If it is surpassed, older
+// items are dropped to make room for new ones.
+var maxQueueSize = 64 * 1024 * 1024 // 64MB; replaced in tests
+
+// sender books payloads for being sent to a given URL by a given client. It uses a
+// size-limited retry queue with a backoff mechanism in case of retriable errors.
+type sender struct {
+	cfg  *senderConfig
+	host string
+
+	wg     sync.WaitGroup // waits for all uploads
+	climit chan struct{}  // acts as a semaphore for limiting concurrent connections
+
+	mu        sync.Mutex  // guards below fields
+	list      *list.List  // send queue
+	size      int         // size of send queue
+	scheduled bool        // reports if a flush is scheduled
+	attempt   int         // retry attempt coming up next
+	timer     *time.Timer // timed flush (triggered by backoff)
+}
+
+// senderConfig specifies the configuration for the sender.
+type senderConfig struct {
+	// client specifies the HTTP client to use when sending requests.
+	client *http.Client
+	// url specifies the URL to send requests too.
+	url *url.URL
+	// apiKey specifies the Datadog API key to use.
+	apiKey string
+	// maxConns specifies the maximum number of allowed concurrent ougoing
+	// connections.
+	maxConns int
+	// recorder specifies the eventRecorder to use when reporting events occurring
+	// in the sender.
+	recorder eventRecorder
+}
+
+// newSender creates a new sender, which uses the given HTTP client, URL and API key
+// to make requests using incoming payloads. The limit of outgoing concurrent connections
+// is defined by climit.
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
+	}
+}
+
+// Push pushes p onto the sender, to be written to the destination.
+func (q *sender) Push(p *payload) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// enqueue the payload
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduled = true
+		q.wg.Add(1)
+		go q.flush()
+	}
```

Comment:
Nit: These 3 things are also done at the end of a `flush`. If we integrated them inside `flush()` itself we could maybe avoid the repetition?

Reply:
I'm not 100% sure I get what you mean here. For the wait group I think it's typical (and makes sense) for it to be outside of the goroutine that it is counting. For the boolean flag, if we integrate it into an asynchronous goroutine, it could cause a race condition in this very function if a new payload comes in before the bool flag is flipped, if that makes sense. That would then potentially cause further goroutines to be created; not a tragedy, but it would be incorrect.

Never the less, I isolated all of this logic out into a new method called `scheduleFlushLocked(delay time.Duration)`. Do you think this is better? It seems nicer to read.

Reply:
Yep I much prefer it with this new `scheduleFlushLocked` :+1: